### DISCUSSION TITLE
Issue with a docker and ollama

### DISCUSSION THREAD

**Discussion by aimagician** [bionic-gpt/bionic-gpt#235] [Q&A]
Here is the output

`bionic-gpt-llm-api-1         | INFO:     172.21.0.9:39886 - "POST /v1/chat/completions HTTP/1.1" 200 OK
bionic-gpt-llm-api-1         | ERROR:    Exception in ASGI application
bionic-gpt-llm-api-1         | Traceback (most recent call last):
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/uvicorn/protocols/http/h11_impl.py", line 408, in run_asgi
bionic-gpt-llm-api-1         |     result = await app(  # type: ignore[func-returns-value]
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 84, in __call__
bionic-gpt-llm-api-1         |     return await self.app(scope, receive, send)
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/fastapi/applications.py", line 1106, in __call__
bionic-gpt-llm-api-1         |     await super().__call__(scope, receive, send)
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/starlette/applications.py", line 122, in __call__
bionic-gpt-llm-api-1         |     await self.middleware_stack(scope, receive, send)
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/starlette/middleware/errors.py", line 184, in __call__
bionic-gpt-llm-api-1         |     raise exc
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/starlette/middleware/errors.py", line 162, in __call__
bionic-gpt-llm-api-1         |     await self.app(scope, receive, _send)
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/starlette/middleware/cors.py", line 83, in __call__
bionic-gpt-llm-api-1         |     await self.app(scope, receive, send)
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
bionic-gpt-llm-api-1         |     raise exc
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
bionic-gpt-llm-api-1         |     await self.app(scope, receive, sender)
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py", line 20, in __call__
bionic-gpt-llm-api-1         |     raise e
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py", line 17, in __call__
bionic-gpt-llm-api-1         |     await self.app(scope, receive, send)
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/starlette/routing.py", line 718, in __call__
bionic-gpt-llm-api-1         |     await route.handle(scope, receive, send)
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/starlette/routing.py", line 276, in handle
bionic-gpt-llm-api-1         |     await self.app(scope, receive, send)
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/starlette/routing.py", line 69, in app
bionic-gpt-llm-api-1         |     await response(scope, receive, send)
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/starlette/responses.py", line 277, in __call__
bionic-gpt-llm-api-1         |     await wrap(partial(self.listen_for_disconnect, receive))
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 597, in __aexit__
bionic-gpt-llm-api-1         |     raise exceptions[0]
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/starlette/responses.py", line 273, in wrap
bionic-gpt-llm-api-1         |     await func()
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/starlette/responses.py", line 262, in stream_response
bionic-gpt-llm-api-1         |     async for chunk in self.body_iterator:
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/litellm/proxy/proxy_server.py", line 707, in async_data_generator
bionic-gpt-llm-api-1         |     async for chunk in response:
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/litellm/main.py", line 204, in _async_streaming
bionic-gpt-llm-api-1         |     raise exception_type(
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/litellm/utils.py", line 4634, in exception_type
bionic-gpt-llm-api-1         |     raise e
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/litellm/utils.py", line 4501, in exception_type
bionic-gpt-llm-api-1         |     raise ServiceUnavailableError(
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/litellm/exceptions.py", line 92, in __init__
bionic-gpt-llm-api-1         |     super().__init__(
bionic-gpt-llm-api-1         |   File "/usr/local/lib/python3.9/site-packages/openai/_exceptions.py", line 79, in __init__
bionic-gpt-llm-api-1         |     super().__init__(message, response.request, body=body)
bionic-gpt-llm-api-1         | AttributeError: 'NoneType' object has no attribute 'request'`

And here is the image:

![Screenshot from 2023-12-22 12-44-00](https://github.com/bionic-gpt/bionic-gpt/assets/1738358/f9cb2389-35ff-4d27-bf0f-5acd6983d632)

Not sure what is wrong? I am using ollama with docker compose, ollama works, i can use it via command line, wanted to check the this gui.

**Comment by kulbinderdio** [comment]
Hi,
Can you just check that the model defined in the litellm section of the ollama.yml file is the same model you have loaded into ollama.

**Comment by aimagician** [comment]
Please take a look at the pictures, of my setup, I am not sure what I am doing wrong.

![Screenshot from 2023-12-22 13-57-09](https://github.com/bionic-gpt/bionic-gpt/assets/1738358/abc18652-1a3e-4e6d-a2f5-7a8b0c8f109d)
![Screenshot from 2023-12-22 13-56-30](https://github.com/bionic-gpt/bionic-gpt/assets/1738358/f1d479fd-c135-4516-ac80-f826c1c0657a)
![Screenshot from 2023-12-22 13-55-35](https://github.com/bionic-gpt/bionic-gpt/assets/1738358/665c5661-669f-46bb-860b-3368ceebb300)
![Screenshot from 2023-12-22 13-55-28](https://github.com/bionic-gpt/bionic-gpt/assets/1738358/5db0b1db-6048-4b7b-af53-e2a687d0d9fc)
![Screenshot from 2023-12-22 13-55-20](https://github.com/bionic-gpt/bionic-gpt/assets/1738358/c365c295-9d18-424f-a4f7-d7467b173ae3)

**Comment by kulbinderdio** [comment]
can you try the following

Update your docker-compose-ollama.yml to be
```
services:
  ollama:
    image: ollama/ollama:latest
    volumes:
      - ./ollama:/root/.ollama
    ports:
      - "11434:11434"
    deploy:
      resources:
        reservations:
          devices:
            - driver: nvidia
              count: 1
              capabilities: [gpu]

  llm-api:
    image: ghcr.io/berriai/litellm:main-v1.10.3
    command: ["/bin/sh", "-c", "pip install async_generator && litellm --model ollama/llama2 --api_base http://ollama:11434 --host 
0.0.0.0 --port 3000"]
    entrypoint: []
    platform: linux/amd64
    ports:
      - "3000:3000"
    extra_hosts:
      - "host.docker.internal:host-gateway"
```

bring the current docker containers down, then start again using
```
docker compose -f docker-compose.yml -f docker-compose-ollama.yml up
```

ensure the model loaded in ollama is the same as specified above
```
docker exec -it bionic_ollama_1 ollama run llama2
```

and then try the front end again

**Comment by aimagician** [comment]
Hello

`Error response from daemon: driver failed programming external connectivity on endpoint bionic-gpt-ollama-1 (d2ca3b179e29229633995905ac92847c38e44ce8739378134f2dfaecf23e42de): Error starting userland proxy: listen tcp4 0.0.0.0:11434: bind: address already in use`

**Comment by kulbinderdio** [comment]
did you shutdown the original ollama docker container you had running?
Shut down all running containers
you can check if any are running by using docker ps

failing that, send me your email and I'll set up a call to try and resolve

**Comment by aimagician** [comment]
Yes, I have shut down all containers, done with the -v flag, so volumes are cleared.

I have done netstat:
`tcp        0      0 127.0.0.1:11434         0.0.0.0:*               LISTEN      3604/ollama   `

`ollama uses the port, changed port, and now I got a message: Error response from daemon: could not select device driver "nvidia" with capabilities: [[gpu]]`

I have nvidia gpu .

**Comment by kulbinderdio** [comment]
Can you check what the GPU count is in the ollama.yml file.
In the original file I posted I had it at 2, but if you have only a single GPU this needs to be 1.

**Comment by aimagician** [comment]
I have used this file to spin up:

`services:
  ollama:
    image: ollama/ollama:latest
    volumes:
      - ./ollama:/root/.ollama
    ports:
      - "11435:11434"
    deploy:
      resources:
        reservations:
          devices:
            - driver: nvidia
              count: 1
              capabilities: [gpu]

  llm-api:
    image: ghcr.io/berriai/litellm:main-v1.10.3
    command: ["/bin/sh", "-c", "pip install async_generator && litellm --model ollama/llama2 --api_base http://ollama:11434 --host 0.0.0.0 --port 3000"]
    entrypoint: []
    platform: linux/amd64
    ports:
      - "3000:3000"
    extra_hosts:
      - "host.docker.internal:host-gateway"
`
The count is set to 1

**Comment by kulbinderdio** [comment]
have you installed the Nvidia container toolkit?

https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html

**Comment by tbendien** [comment]
@aimagician take a look at the runbook on my blog. What OS are you using? I am running on Ubuntu 22.04.
https://gtcyberlabs.org/f/install-bionic-gpt-on-an-ubuntu-bare-metal-host-with-a-nvidia-gpu
